### PR TITLE
KAFKA-17035 Add debug log to retention cleanupLogs method to help troubleshoot issues

### PR DESCRIPTION
**Pull request by brandboat**
related to https://issues.apache.org/jira/browse/KAFKA-17035,

add more debug logs in UnifiedLog.scala

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/main/scala/kafka/log/UnifiedLog.scala [hunk 1]
```diff
@@ -1587,20 +1590,23 @@ class UnifiedLog(@volatile var logStartOffset: Long,
     if (retentionSize < 0 || size < retentionSize) return 0
     var diff = size - retentionSize
     def shouldDelete(segment: LogSegment, nextSegmentOpt: Option[LogSegment]): Boolean = {
-      if (diff - segment.size >= 0) {
+      val shouldDelete = diff - segment.size >= 0
```
Comment (chia7712) [inline core/src/main/scala/kafka/log/UnifiedLog.scala]: Should we cache the `segment.size`?

File: core/src/main/scala/kafka/log/UnifiedLog.scala [hunk 2]
```diff
@@ -1587,20 +1590,23 @@ class UnifiedLog(@volatile var logStartOffset: Long,
...
 
     deleteOldSegments(shouldDelete, RetentionSizeBreach(this, remoteLogEnabled()))
   }
 
   private def deleteLogStartOffsetBreachedSegments(): Int = {
     def shouldDelete(segment: LogSegment, nextSegmentOpt: Option[LogSegment]): Boolean = {
-      nextSegmentOpt.exists(_.baseOffset <= (if (remoteLogEnabled()) localLogStartOffset() else logStartOffset))
+      val shouldDelete = nextSegmentOpt.exists(_.baseOffset <= (if (remoteLogEnabled()) localLogStartOffset() else logStartOffset))
+      debug(s"$segment logStartOffset breached: $shouldDelete, nextSegmentOpt=$nextSegmentOpt, " +
+        s"${if (remoteLogEnabled()) s"localLogStartOffset=${localLogStartOffset()}" else s"logStartOffset=$logStartOffset"}")
```
Comment (chia7712) [inline core/src/main/scala/kafka/log/UnifiedLog.scala]: ditto for `localLogStartOffset`

**Review by chia7712 — COMMENTED** [review]
@brandboat thanks for the nice debugging message

File: core/src/main/scala/kafka/log/UnifiedLog.scala (hunk 2 above)
Comment (brandboat) [reply core/src/main/scala/kafka/log/UnifiedLog.scala]: Thanks for the review, I've addressed everything you mentioned.

File: core/src/main/scala/kafka/log/UnifiedLog.scala [hunk 3]
```diff
@@ -1576,7 +1576,10 @@ class UnifiedLog(@volatile var logStartOffset: Long,
     val startMs = time.milliseconds
 
     def shouldDelete(segment: LogSegment, nextSegmentOpt: Option[LogSegment]): Boolean = {
-      startMs - segment.largestTimestamp > retentionMs
+      val largestTimestamp = segment.largestTimestamp
+      val shouldDelete = startMs - largestTimestamp > retentionMs
+      debug(s"$segment retentionMs breached: $shouldDelete, startMs=$startMs, largestTimestamp=$largestTimestamp, retentionMs=$retentionMs")
```
Comment (chia7712) [inline core/src/main/scala/kafka/log/UnifiedLog.scala]: `LogSegment#toString` already include `lastModifiedTime` and `largestRecordTimestamp`. Maybe we can delete `largestTimestamp` as it is redundant ?

**Review by chia7712 — COMMENTED** [review]
@brandboat thanks for updates. one last comment PTAL

File: core/src/main/scala/kafka/log/UnifiedLog.scala (hunk 3 above)
Comment (brandboat) [reply core/src/main/scala/kafka/log/UnifiedLog.scala]: You are right, thanks.

**Comment by chia7712** [comment]
@brandboat could you please rebase code to have CI again?

**Comment by chia7712** [comment]
the failed test is traced by https://issues.apache.org/jira/browse/KAFKA-17309. others pass on my local. will merge it

**Pull request merged**